### PR TITLE
opt: fold constants during normalization

### PR DESCRIPTION
**Pull request by rytaft**
This commit adds two normalization rules:
`FoldBinary` and `FoldUnary`. These rules are used to fold constants
as part of the optimizer's normalization phase.

`FoldBinary` evaluates a binary operation over constant inputs,
replacing the entire operation with a constant. The rule applies
as long as the evaluation would not cause an error. Any errors should
be saved for execution time, since it's possible that the given
operation will not be executed. For example:
  `SELECT CASE WHEN true THEN 42 ELSE 1/0 END`

`FoldUnary` is similar to `FoldBinary`, but it involves a unary operation
over a single constant input.

As part of this change, convert the `BinOps` and `UnaryOps` slices of
builtin functions to contain `*tree.BinOp` and `*tree.UnaryOp` rather than
`tree.BinOp` and `tree.UnaryOp`. This does not cause any extra allocations,
and allows us to avoid copying around these large objects.

Release note: None

### CODE REVIEW

**Review by andy-kimball — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29679)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/memo/typing.go, line 339 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-LLlQwgoBs-0S8XxzO3H:-LLlQkdD-773P4jKkngk:b-vmbbll) ([raw file](https://github.com/cockroachdb/cockroach/blob/db085f9c0e96df3bf8d7dfb71e4550990b404984/pkg/sql/opt/memo/typing.go#L339)):*
> ```Go
> // found, FindBinaryOverload returns true, plus the overload itself.
> // If an overload is not found, FindBinaryOverload returns false.
> func FindBinaryOverload(op opt.Operator, leftType, rightType types.T) (_ tree.BinOp, ok bool) {
> ```

`tree.BinOp` is a huge struct (88 bytes), and shouldn't be passed by value (very unfortunate original design). That's the reason for creating the `overload` type and having a separate `EvalUnaryOp` method. I don't know of any way to take its address, since it lives in a `map`. So I settled for copying it to stack just once (assigning to temporary variable `o`).
___
*[pkg/sql/opt/norm/rules/fold_constants.opt, line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-LLlSWAS1XesX_osQNBa:-LLlSWAS1XesX_osQNBb:byu648z) ([raw file](https://github.com/cockroachdb/cockroach/blob/db085f9c0e96df3bf8d7dfb71e4550990b404984/pkg/sql/opt/norm/rules/fold_constants.opt#L19)):*
> ```
> (Binary
>   $left:(Const)
>   $right:(Const) & (CanFoldBinary (OpName) $left $right)
> ```

Optgen now allows this pattern:
```
(Binary
  $left:(Const)
  $right:(Const) & (FoldSucceeded $result:(FoldBinary (OpName) $left $right))
)
=>
$result
 ```
Not sure how much this matters, but seems like easy perf win for small degradation in clarity. `FoldBinary` would return group=0 if it failed, and `FoldSucceeded` would return false if its input = 0.
___
*[pkg/sql/sem/tree/eval.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-LLlPCgjCb1GMhlB2Ul3:-LLlPCgk-1xD4QeLwvVq:bvwrgfc) ([raw file](https://github.com/cockroachdb/cockroach/blob/db085f9c0e96df3bf8d7dfb71e4550990b404984/pkg/sql/sem/tree/eval.go#L73)):*
> ```Go
> 	Typ          types.T
> 	ReturnType   types.T
> 	NeedsContext bool
> ```

We don't need this for binary/unary operators. The `PREPARE` code detects changes to:
1. Location (timezone)
2. Current database
3. Current search path

The only reason binary/unary operators use context is for location/timezone and for resolving placeholders. Location/timezone is important enough that we should still be able to constant fold affected expressions, especially since it virtually never changes.

So you can get rid of this. We will need something like this for Functions. However, in that case we should generalize `Impure` to be an enumeration rather than boolean.

**Review by rytaft — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29679)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/memo/typing.go, line 339 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-LLlQwgoBs-0S8XxzO3H:-LLo4K3qEjAvVliVnVnP:b8cdysa) ([raw file](https://github.com/cockroachdb/cockroach/blob/db085f9c0e96df3bf8d7dfb71e4550990b404984/pkg/sql/opt/memo/typing.go#L339)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

`tree.BinOp` is a huge struct (88 bytes), and shouldn't be passed by value (very unfortunate original design). That's the reason for creating the `overload` type and having a separate `EvalUnaryOp` method. I don't know of any way to take its address, since it lives in a `map`. So I settled for copying it to stack just once (assigning to temporary variable `o`).
</blockquote></details>

Added a new function `findBinaryOverloadIdx` to avoid copying it altogether. Does this seem better?
___
*[pkg/sql/opt/norm/rules/fold_constants.opt, line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-LLlSWAS1XesX_osQNBa:-LLo51PTBw9XpRgLF068:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db085f9c0e96df3bf8d7dfb71e4550990b404984/pkg/sql/opt/norm/rules/fold_constants.opt#L19)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Optgen now allows this pattern:
```
(Binary
  $left:(Const)
  $right:(Const) & (FoldSucceeded $result:(FoldBinary (OpName) $left $right))
)
=>
$result
 ```
Not sure how much this matters, but seems like easy perf win for small degradation in clarity. `FoldBinary` would return group=0 if it failed, and `FoldSucceeded` would return false if its input = 0.
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/eval.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-LLlPCgjCb1GMhlB2Ul3:-LLo55G704eznmdvlvay:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db085f9c0e96df3bf8d7dfb71e4550990b404984/pkg/sql/sem/tree/eval.go#L73)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

We don't need this for binary/unary operators. The `PREPARE` code detects changes to:
1. Location (timezone)
2. Current database
3. Current search path

The only reason binary/unary operators use context is for location/timezone and for resolving placeholders. Location/timezone is important enough that we should still be able to constant fold affected expressions, especially since it virtually never changes.

So you can get rid of this. We will need something like this for Functions. However, in that case we should generalize `Impure` to be an enumeration rather than boolean.
</blockquote></details>

Done.

**Review by andy-kimball — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29679)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/memo/typing.go, line 339 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-LLlQwgoBs-0S8XxzO3H:-LLo8jUfB4l4GoKXNm16:b30woet) ([raw file](https://github.com/cockroachdb/cockroach/blob/db085f9c0e96df3bf8d7dfb71e4550990b404984/pkg/sql/opt/memo/typing.go#L339)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Added a new function `findBinaryOverloadIdx` to avoid copying it altogether. Does this seem better?
</blockquote></details>

Possibly; hard to know since now we're doing extra map lookups in its place.  I think the best fix would be to export `binOpOverload` (probably as `BinOpOverloads`, so you can just take the address of a slice entry.

**Review by rytaft — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29679)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/memo/typing.go, line 339 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-LLlQwgoBs-0S8XxzO3H:-LLoN-HV6l8HW4r6lM72:b-76f6zy) ([raw file](https://github.com/cockroachdb/cockroach/blob/db085f9c0e96df3bf8d7dfb71e4550990b404984/pkg/sql/opt/memo/typing.go#L339)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Possibly; hard to know since now we're doing extra map lookups in its place.  I think the best fix would be to export `binOpOverload` (probably as `BinOpOverloads`, so you can just take the address of a slice entry.
</blockquote></details>

Per our offline discussion, changed `tree.BinOps` to contain `*tree.BinOp` instead of `tree.BinOp`.

**Review by andy-kimball — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29679)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/eval.go, line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-LLoQqw43JxHFpgq1S04:-LLoQqw43JxHFpgq1S05:bm8x16j) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dd3b18c9b4b5de6d777eef7f36dfe842c7a0904/pkg/sql/sem/tree/eval.go#L293)):*
> ```Go
> 	for _, t := range types.AnyNonArray {
> 		typ := t
> 		BinOps[Concat] = append(BinOps[Concat], BinOp{
> ```

You need to use `&BinOp` here as well. Make sure you find any other places as well.

File: pkg/sql/opt/norm/rules/fold_constants.opt [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# =============================================================================
+# fold_consants.opt contains normalization rules for folding constants.
```
Comment (justinj) [inline pkg/sql/opt/norm/rules/fold_constants.opt]: nit: `fold_constants.opt`

**Review by rytaft — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29679)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/norm/rules/fold_constants.opt, line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-LLkn13QQlLgb8rXxvZ3-r2-2:-LLoXWgi5scD21sGpmv_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dd3b18c9b4b5de6d777eef7f36dfe842c7a0904/pkg/sql/opt/norm/rules/fold_constants.opt#L2)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: `fold_constants.opt`
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/eval.go, line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-LLoQqw43JxHFpgq1S04:-LLoX_Kk36nhn6I4VpKj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dd3b18c9b4b5de6d777eef7f36dfe842c7a0904/pkg/sql/sem/tree/eval.go#L293)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

You need to use `&BinOp` here as well. Make sure you find any other places as well.
</blockquote></details>

Done.

**Review by andy-kimball — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-:-LLo_8B54axsa0CN1pL3:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Review by andy-kimball — APPROVED** [review]
Nice work.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-:-LLo_DMYDjHUmtFUMkMW:bdtx8rj)* status: :shipit: complete! 1 of 0 LGTMs obtained

**Review by rytaft — COMMENTED** [review]
Thanks! As we discussed offline, @andy-kimball, I made one more minor change to get rid of the `Eval` functions in `memo/typing.go` and instead export `FindBinaryOverload` and `FindUnaryOverload`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-:-LLogg5J1wb-tp_laP2Y:b-z1w8ng)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Review by rytaft — COMMENTED** [review]
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-:-LLop-an80rwiH7E89rI:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Review by RaduBerinde — COMMENTED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-:-LLq2cYcEs5fM4jiLg_x:bg7igkv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/sql/opt/norm/rules/fold_constants.opt, line 18 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-LLq2Atj0thF_2zHIxIt:-LLq2Atj0thF_2zHIxIu:bqegjuh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c81e25e7438eb140d002f4b295e39c02b3d17c8/pkg/sql/opt/norm/rules/fold_constants.opt#L18)):*
> ```
> (Binary
>   $left:(Const)
>   $right:(Const) & (FoldSucceeded $result:(FoldBinary (OpName) $left $right))
> ```

The way these rules are written is really cool!

**Comment by rytaft** [comment]
___
*[pkg/sql/opt/norm/rules/fold_constants.opt, line 18 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/29679#-LLq2Atj0thF_2zHIxIt:-LM31Vpf89Jk24OqdinX:b19caoo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c81e25e7438eb140d002f4b295e39c02b3d17c8/pkg/sql/opt/norm/rules/fold_constants.opt#L18)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

The way these rules are written is really cool!
</blockquote></details>

Thanks! Credit to @andy-kimball

**Pull request merged**